pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -497,46 +586,84 @@ public void close() {
-                if (flush) {
-                    cnx.ctx().writeAndFlush(cmd, cnx.ctx().voidPromise());
+                if (ackResponseEnabled) {
+                    long requestId = consumer.getClient().newRequestId();
+                    ByteBuf cmd = Commands.newMultiMessageAck(consumer.consumerId, entriesToAck, requestId);
+                    completableFuture = cnx.newAckForResponse(cmd, requestId);
                 } else {
-                    cnx.ctx().write(cmd, cnx.ctx().voidPromise());
+                    ByteBuf cmd = Commands.newMultiMessageAck(consumer.consumerId, entriesToAck, -1);
+                    cnx.ctx().writeAndFlush(cmd, cnx.ctx().voidPromise());
+                    completableFuture = CompletableFuture.completedFuture(null);
                 }
             } else {
+                // if don't support multi message ack, it also support ack response, so we should not think about the
+                // ack response in this logic
                 for (MessageIdImpl cMsgId : chunkMsgIds) {
                     ByteBuf cmd = Commands.newAck(consumerId, cMsgId.getLedgerId(), cMsgId.getEntryId(),
-                            lastCumulativeAckSet, ackType, validationError, map);
-                    if (flush) {
-                        cnx.ctx().writeAndFlush(cmd, cnx.ctx().voidPromise());
-                    } else {
-                        cnx.ctx().write(cmd, cnx.ctx().voidPromise());
-                    }
+                            bitSet, ackType, null, map, -1);
+                    cnx.ctx().writeAndFlush(cmd, cnx.ctx().voidPromise());
                 }
+                completableFuture = CompletableFuture.completedFuture(null);
             }
-            this.consumer.unAckedChunkedMessageIdSequenceMap.remove(msgId);
         } else {
-            ByteBuf cmd = Commands.newAck(consumerId, msgId.getLedgerId(), msgId.getEntryId(), lastCumulativeAckSet,
-                    ackType, validationError, map, txnidLeastBits, txnidMostBits, -1);
-            if (flush) {
-                cnx.ctx().writeAndFlush(cmd, cnx.ctx().voidPromise());
+            if (ackResponseEnabled) {
+                long requestId = consumer.getClient().newRequestId();
+                ByteBuf cmd = Commands.newAck(consumerId, msgId.getLedgerId(), msgId.getEntryId(), bitSet,
+                        ackType, null, map, requestId);
+                completableFuture = cnx.newAckForResponse(cmd, requestId);
             } else {
-                cnx.ctx().write(cmd, cnx.ctx().voidPromise());
+                ByteBuf cmd = Commands.newAck(consumerId, msgId.getLedgerId(), msgId.getEntryId(), bitSet,
+                        ackType, null, map, -1);
+                cnx.ctx().writeAndFlush(cmd, cnx.ctx().voidPromise());
+                completableFuture = CompletableFuture.completedFuture(null);
+            }
+        }
+        return completableFuture;
```

Comment:
I think we can use a method `getRequestId()` and a method `getCompletableFuture()`. I think this will make the logic simpler